### PR TITLE
Error estimation with variances from observations

### PR DESCRIPTION
**Pull request by marcusps**
Could you take a peek, @blakejohnson ?  

I am not happy to have the `pinv` for computing the covariance of the estimates, but it seemed necessary.  I tried inverses of symmetric and Hermitian matrices, but they still failed occasionally on some test data I have (fits to Ramsey interferometry data). Any suggestions would be welcome.

The computation of the covariance matrix in terms of the variance of the observations is based on a similar calculation for weighted linear least squares (in essence, we assume linearization around the minimum found).

### CODE REVIEW

**Comment by marcusps** [comment]
I remove the pseudo inverse.  It was only needed in a very pathological case using real world data (our of more than a hundred).  Ignoring that single example, all other cases get by just fine with the usual inverse.

File: src/curve_fit.jl [hunk 1]
```diff
@@ -88,3 +97,22 @@ function estimate_errors(fit::LsqFitResult, alpha=0.95)
 	dist = TDist(fit.dof)
 	std_error *= quantile(dist, alpha)
 end
+
+function estimate_errors(fit::LsqFitResult, vars::Vector{Float64}, alpha=0.95; atol=1e-10)
+    # computes (1-alpha) error estimates from
+    #   fit   : a LsqFitResult from a curve_fit()
+    #   alpha : alpha percent confidence interval, (e.g. alpha=0.95 for 95% CI)
```
Comment (blakejohnson) [inline src/curve_fit.jl]: need to document the `vars` input.

Comment (blakejohnson) [inline src/curve_fit.jl]: and also `atol`

File: src/curve_fit.jl [hunk 2]
```diff
@@ -88,3 +97,22 @@ function estimate_errors(fit::LsqFitResult, alpha=0.95)
...
+    covar = estimate_covar(fit, vars)
+    
+    # then the standard errors are given by the sqrt of the diagonal
+    vars = diag(covar)
+    vratio = minimum(vars)/maximum(vars) 
+    if !isapprox(vratio, 0.0, atol=atol) && vratio < 0.0
+        error("Covariance matrix is negative")
+    end
+    std_error = sqrt(abs(vars))
```
Comment (blakejohnson) [inline src/curve_fit.jl]: is the `abs` just to deal with some corner case? or are there expected scenarios where the diagonal of the covariance matrix has negative entries?

File: test/curve_fit.jl [hunk 3]
```diff
@@ -14,4 +14,18 @@ let
...
+    yvars = 1e-4*randn(length(xdata))
+    ydata = model(xdata, [1.0, 2.0]) + sqrt(yvars).*randn(length(xdata))
+
+    fit = curve_fit(model, xdata, ydata, 1./sqrt(yvars), [0.5, 0.5])
+    println("norm(fit.param - [1.0, 2.0]) < 0.05 ? ", norm(fit.param - [1.0, 2.0]))
+    @assert norm(fit.param - [1.0, 2.0]) < 0.05
+    @test fit.converged
+
+    # can also get error estimates on the fit parameters
+    errors = estimate_errors(fit,yvars)
```
Comment (blakejohnson) [inline test/curve_fit.jl]: does it feel right to you that the interface for `estimate_errors` changes when the data points are not weighted evenly? the alternative would be to store enough information in `LsqFitResult`.

Comment (marcusps) [inline test/curve_fit.jl]: Good point.  Putting the variance information in `LsqFitResult` is probably better.

File: src/curve_fit.jl (hunk 2 above)
Comment (marcusps) [inline src/curve_fit.jl]: I think up to numerical precision it is possible it is negative, as long as it is approximately zero — mathematically, of course, the construction should ensure positive semidefinites.

In my tests with the Ramsey data negativity does seem to come up, if I remember correctly, but adding a tolerance around 0.0 (as implied by the `isapprox` test) deals with that.

**Comment by marcusps** [comment]
I think the interface is cleaner now, @blakejohnson, and everything documented in the README.
